### PR TITLE
Fast sync (WIP)

### PR DESCRIPTION
**Pull request by jomuel**
Partly Implements #153

- creates `consensus/fast-sync` and `FastSyncService`
- Implements `FastSyncRequest::ValidateHashes`
- a tool to generate the hashes of hashes from a synced blockchain

Collaborative effort with @dllud

The missing `FastSyncRequest::ValidateBlock` will be implemented  in a follow-up PR

### CODE REVIEW

File: consensus/fast-sync/src/fast_sync.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    fn call(&mut self, req: FastSyncRequest) -> Self::Future {
+        Box::pin(async move {
+            match req {
+                FastSyncRequest::ValidateHashes {
+                    start_height,
+                    block_ids,
+                } => validate_hashes(start_height, &block_ids).await,
+            }
+        })
+    }
```
Comment (SyntheticBird45) [inline consensus/fast-sync/src/fast_sync.rs:109]:
```suggestion
    fn call(&mut self, FastSyncRequest::ValidateHashes { start_height, block_ids }: FastSyncRequest) -> Self::Future {
        Box::pin(async move {
            validate_hashes(start_height, &block_ids).await
        })
    }
```
@Boog900

**Review by SyntheticBird45 — APPROVED** [review]

File: consensus/fast-sync/src/fast_sync.rs (hunk 1 above)
Comment (Boog900) [reply consensus/fast-sync/src/fast_sync.rs:109]: There are going to be other requests + I don't really like destructuring in the function arguments

File: consensus/fast-sync/src/fast_sync.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+use hex_literal::hex;
+
+use std::{
+    cmp,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::Service;
+
```
Comment (Boog900) [inline consensus/fast-sync/src/fast_sync.rs:13]: Same here with the imports:

```suggestion
use std::{
    cmp,
    future::Future,
    pin::Pin,
    task::{Context, Poll},
};

#[allow(unused_imports)]
use hex_literal::hex;
use tower::Service;

use crate::{hash_of_hashes, BlockId, HashOfHashes};
```

**Review by Boog900 — CHANGES_REQUESTED** [review]
Thanks for working on this guys!

File: consensus/fast-sync/src/fast_sync.rs (hunk 1 above)
Comment (jomuel) [reply consensus/fast-sync/src/fast_sync.rs:109]: We are about to implement the second enum variant, `FastSyncRequest::ValidateBlock`, in the follow-up pr. So this change would just get undone then.

File: consensus/fast-sync/src/create.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        write!(&mut s, "\thex!(\"").unwrap();
+        for byte in hash {
+            write!(&mut s, "{:02x}", byte).unwrap();
+        }
+        writeln!(&mut s, "\"),").unwrap();
+    }
+
+    writeln!(&mut s, "]").unwrap();
+
+    s
```
[the remark this answers is not in the document]
Comment (jomuel) [reply consensus/fast-sync/src/create.rs]: Yes, done.

**Review by Boog900 — APPROVED** [review]

**Pull request merged**